### DISCUSSION TITLE
Cuda async template unavailable

### DISCUSSION THREAD

**Discussion by bakpaul** [sofa-framework/sofa#3644] [Write / Run a simulation] [answered]
**This is a template instantiation problem when running a simulation.** 

Trigger warning : this happens in a private plugin, but it seems to be really related to sofa because it worked in last November but now it doesn't work anymore (I haven't used this scene since November so I don't have more precision on when it stopped working) and I am working in the master branch, updating each day. 

**To sum up a bit the context  :** 

We use a private plugin for GPU accelerated LDL decomposition (SofaCUDASolver, some of you have access to it on the Inira Gitlab). In this plugin, we are instantiating an object with three template in the following fashion :

`CudaSparseLDLSolver< CompressedRowSparseMatrix<type::Mat<3,3,float> >,CudaBaseVector<float>, NoThreadManager >`
or 
`CudaSparseLDLSolver< CompressedRowSparseMatrix<type::Mat<3,3,float> >,CudaBaseVector<float>, AsyncThreadManager >`

Some other instantiations exists with other matrices, but what is important to note here is that for every tuple containing  `NoThreadManager` its equivalent containing `AsyncThreadManager` exists. 

**Now the problem  :**

I have a scene, with a CudaVec3f mechanical object on which I am trying to use the CudaSparseLDLSolver with the following template `<CudaSparseLDLSolver name="precond" template="CompressedRowSparseMatrixMat3x3,CudaBaseVectorf,Async" />`
but sofa throws an error which is :
`[ERROR]   [CudaSparseLDLSolver(precond)] Requested template 'CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Async' is not compatible with the current context. Falling back to the first compatible template: 'CompressedRowSparseMatrixMat3x3f'.` although the exact same scene worked perfectly at one point. 

So you are wondering what is this single template ? When I list the available templates I've got :
` [CompressedRowSparseMatrixMat3x3f, CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Async, CompressedRowSparseMatrixf, CompressedRowSparseMatrixf,CudaBaseVectorf,Async, EigenBaseSparseMatrixf, EigenBaseSparseMatrixf,CudaBaseVectorf,Async] `
The single template seems to be the `NoThreadManager` tuples and the rest are the ones I want to use. But currently I cannot instantiate any of the async tuple in my scene which causes serious slow down...

Can you help me ?

**Comment by alxbilger** [comment]
I suspect the problem comes from this commit: https://github.com/sofa-framework/sofa/pull/3592

Your scene uses a template alias (`CompressedRowSparseMatrixMat3x3`). By default, this alias corresponds to `CompressedRowSparseMatrixMat3x3d` (because `SReal` is double). It would correspond to `CompressedRowSparseMatrixMat3x3f` if you compile SOFA with `SReal` a `float`.

Therefore, you are implicitly asking the full template `CompressedRowSparseMatrixMat3x3d,CudaBaseVectorf,Async`, but you want to ask `CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Async`. So the solution is simply to add the `f` suffix to `CompressedRowSparseMatrixMat3x3`.

I hope it solves your problem

**Reply by bakpaul** [reply]
Thank you for your answer.

Sorry but this copy/past was the result of many attempt into giving the right template. At first it was indeed `<CudaSparseLDLSolver name="precond" template="CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Async" />` but the same error is prompted : `'CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Async' is not compatible with the current context. Falling back to the first compatible template: 'CompressedRowSparseMatrixMat3x3f'.`

So, no it doesn't solve the problem...

**Comment by bakpaul** [comment]
To add info :

In the class header, there is a SOFA_CLASS declaration with the following parameters `SOFA_CLASS(SOFA_TEMPLATE3(CudaSparseLDLSolver,TMatrix,TVector,TThreadManager),SOFA_TEMPLATE3(sofa::component::linearsolver::direct::SparseLDLSolverImpl,TMatrix,TVector,TThreadManager));` 

And the class doesn't implement any static method (no `HasGetCustomTemplateName` or `templateName`) so normally, the two sets of tuples should produce the same template structure (during the call to the method `getOverridableTemplateName()` in BaseClassNameHelper.h, using the `GetDefaultTemplateName` added by the macro, when adding the object to the factory) : a triplet with only one difference being the last template. Here, even the second template is missing for the `NoThread` template, which is strange.

**Comment by bakpaul** [comment]
More info (I'm trying to solve this in the meantime), the difference in the template name came from the fact that `CudaSparseLDLSolver` inherited from `MatrixLinearSolver` which has a specialization for each of the two `ThreadManager` type : `NoThread`and `Async`. And the Sync one had a static `GetCustomTemplateName` method (the traits are passed through inheritance). So I've removed it just to test again with consistent template names (and I modified the `NoThreadManager::Name()` method, to output a "Sync"). Thanks to this I have a consistent template list : 
`[CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Async, CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Sync, CompressedRowSparseMatrixf,CudaBaseVectorf,Async, CompressedRowSparseMatrixf,CudaBaseVectorf,Sync, EigenBaseSparseMatrixf,CudaBaseVectorf,Async, EigenBaseSparseMatrixf,CudaBaseVectorf,Sync]`

But now it gets more interesting : guess what happens when I try to put this as template 
`<CudaSparseLDLSolver name="precond" template="CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Sync" />` 
....I get this error : 
`[ERROR]   [CudaSparseLDLSolver(precond)] Requested template 'CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Sync' is not compatible with the current context. Falling back to the first compatible template: 'CompressedRowSparseMatrixMat3x3f,CudaBaseVectorf,Async'.` !!!!!
Now the template that I originally wanted is the "First compatible", but the previous "First compatible" is not compatible anymore ! 

I'll try to go deeper to see where it is broken...

**Comment by bakpaul** [comment, marked as answer]
That's it ! I found the problem : it comes from the automation of the template declaration regarding the type of SReal, so the commit you suggested @alxbilger is responsible but indirectly. 
The problem comes from the fact that you are replacing automatically the deprecated template at the begining of the `createObject` method of the factory, but even if we compile sofa in Double, the CUDA template may be compiled in Float (which is the case here for me).  So the problem was that it replaced the `CompressedRowSparseMatrixMat3x3f` automatically by `CompressedRowSparseMatrixMat3x3d` resulting in the following template name `CompressedRowSparseMatrixMat3x3d,CudaBaseVectorf,Sync` which is in fact not in the factory, and then the first compatible on was chosen. 

So this should either not be tested on Cuda object, which seems tricky or this template deprecation should be removed for the matrices. 

Should I create an issue ?

**Reply by alxbilger** [reply]
Oh ok I understand. Yes Cuda does not behave the same way than the rest of SOFA, so the automatic replacement of floating-point type cannot apply. Actually, we were even lucky before this commit that we did not have more problems. It comes from the fact that Cuda types are specific, not the SOFA ones. But here, we use the same type for the matrices.

I suggest to remove the lines
```cpp
static RegisterTemplateAlias CompressedRowSparseMatrixfAlias("CompressedRowSparseMatrixf", sofa::linearalgebra::CompressedRowSparseMatrix<SReal>::Name(), isSRealDouble());
static RegisterTemplateAlias CompressedRowSparseMatrixMat3x3fAlias("CompressedRowSparseMatrixMat3x3f", sofa::linearalgebra::CompressedRowSparseMatrix<type::Mat<3, 3, SReal>>::Name(), isSRealDouble());
```
and
```cpp
static RegisterTemplateAlias CompressedRowSparseMatrixdAlias("CompressedRowSparseMatrixd", sofa::linearalgebra::CompressedRowSparseMatrix<SReal>::Name(), isSRealFloat());
static RegisterTemplateAlias CompressedRowSparseMatrixMat3x3dAlias("CompressedRowSparseMatrixMat3x3d", sofa::linearalgebra::CompressedRowSparseMatrix<type::Mat<3, 3, SReal>>::Name(), isSRealFloat());
```

Can you submit a pull request mentioning this issue? Thanks.

**Reply by alxbilger** [reply]
For the records: https://github.com/sofa-framework/sofa/pull/3646

**Comment by yangwxb** [comment]


**Reply by bakpaul** [reply]
There are two main problem with you scene : 
1. Your Surface node misses a MechanicalObject. You need one in every node except if the node contains a OglLModel, which replaces the MechanicalObject.
2. Again, The Surface node is missing a mechanical mapping. You should specify it, I suggest to use an IdentityMapping given the fact that you are using a TopologicalMapping (which by the way is not a mechanical mapping, just a topological one).

**Reply by yangwxb** [reply]
thanks @bakpaul ,your first tip worked, now there are no errors and it works.But I am not sure whether it is also effective for other scenarios, I still need to learn the basic principle, may disturb you later, very sorry. 
Good luck with your work!

**Reply by bakpaul** [reply]
Hello, I would prefere if you started a new thread. This is not how Github discussion are intended to work.

**Reply by yangwxb** [reply]
Okay, I'm sorry
